File: backend/src/test/java/vn/edu/uit/careerhub/auth/TokenServiceTest.java
package vn.edu.uit.careerhub.auth;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import java.util.UUID;

import org.junit.jupiter.api.Test;

import vn.edu.uit.careerhub.common.AppException;
import vn.edu.uit.careerhub.config.AppProperties;

class TokenServiceTest {
    @Test void signsAndVerifiesStudentContext(){TokenService service=new TokenService(properties("0123456789abcdefghijklmnopqrstuvwxyz"));UUID userId=UUID.randomUUID(),studentId=UUID.randomUUID();AuthUser user=new AuthUser(userId,"student@student.uit.edu.vn",UserRole.STUDENT,UserStatus.ACTIVE,null,0,null,"Student","UIT",studentId,null);var token=service.signAccessToken(user);var principal=service.verifyAccessToken(token.accessToken());assertThat(principal.userId()).isEqualTo(userId);assertThat(principal.studentProfileId()).isEqualTo(studentId);assertThat(principal.role()).isEqualTo(UserRole.STUDENT);assertThat(token.expiresIn()).isEqualTo(900);}
    @Test void rejectsTamperedToken(){TokenService service=new TokenService(properties("0123456789abcdefghijklmnopqrstuvwxyz"));AuthUser user=new AuthUser(UUID.randomUUID(),"admin@uit.edu.vn",UserRole.UIT_ADMIN,UserStatus.ACTIVE,null,0,null,"Admin","UIT",null,null);String token=service.signAccessToken(user).accessToken();assertThatThrownBy(()->service.verifyAccessToken(token.substring(0,token.length()-2)+"aa")).isInstanceOf(AppException.class);}
    @Test void createsOpaqueRefreshTokensAndStableHashes(){TokenService service=new TokenService(properties("0123456789abcdefghijklmnopqrstuvwxyz"));String first=service.createRefreshToken(),second=service.createRefreshToken();assertThat(first).hasSize(64).isNotEqualTo(second);assertThat(service.hashOpaqueToken(first)).hasSize(64).isEqualTo(service.hashOpaqueToken(first));}
    @Test void rejectsWeakSigningSecret(){assertThatThrownBy(()->new TokenService(properties("too-short"))).isInstanceOf(IllegalStateException.class);}
    private AppProperties properties(String secret){return new AppProperties("test","http://localhost:5173","http://localhost:5173",new AppProperties.Database("postgresql://localhost/test","",5),new AppProperties.Security(secret,"issuer","audience",900,7,"refresh",false,"lax","student.uit.edu.vn"),new AppProperties.Email(false,"","",10,5),new AppProperties.Storage(false,"","","","bucket",600,300),"cron-secret");}
}

File: backend-express-legacy/src/modules/auth/token.service.test.ts
import { describe, expect, it } from "vitest";

import type { AuthUser } from "./auth.types.js";
import { TokenService } from "./token.service.js";

const user: AuthUser = {
  id: "00000000-0000-4000-8000-000000000011",
  email: "20521067@student.uit.edu.vn",
  role: "STUDENT",
  status: "ACTIVE",
  passwordHash: "not-returned-in-token",
  failedLoginAttempts: 0,
  lockedUntil: null,
  displayName: "Nguyễn Minh Khoa",
  organization: "20521067",
  studentProfileId: "00000000-0000-4000-8000-000000002001",
  companyId: null,
};

describe("TokenService", () => {
  it("should_sign_and_verify_access_token_with_scope", async () => {
    const service = new TokenService({
      secret: "test-secret-with-at-least-32-characters",
      issuer: "test-issuer",
      audience: "test-audience",
      accessTokenTtlSeconds: 900,
    });

    const signed = await service.signAccessToken(user);
    const principal = await service.verifyAccessToken(signed.accessToken);

    expect(signed.expiresIn).toBe(900);
    expect(principal).toMatchObject({
      userId: user.id,
      email: user.email,
      role: "STUDENT",
      studentProfileId: user.studentProfileId,
      companyId: null,
    });
  });

  it("should_hash_refresh_tokens_without_storing_the_plain_value", () => {
    const service = new TokenService({
      secret: "test-secret-with-at-least-32-characters",
      issuer: "test-issuer",
      audience: "test-audience",
      accessTokenTtlSeconds: 900,
    });
    const token = service.createRefreshToken();
    const hash = service.hashOpaqueToken(token);

    expect(token.length).toBeGreaterThanOrEqual(64);
    expect(hash).toHaveLength(64);
    expect(hash).not.toContain(token);
  });
});
